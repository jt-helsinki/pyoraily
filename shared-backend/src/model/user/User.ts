/*
 *
 * MIT License.
 *
 */
import { UserRole } from './UserRole';
import { UserStatus } from './UserStatus';

export enum Gender {
  MALE = 'male',
  FEMALE = 'female',
  NA = 'na',
}

export interface User {
  id: string;
  firstName: string;
  lastName: string;
  fullName?: string;
  yearOfBirth?: number;
  gender?: Gender;
  email: string;
  uciId?: number;
  status: UserStatus;
  createdAt: string;
  updatedAt?: string;
  lastLogin?: string;
  userRoles: UserRole[];
}
