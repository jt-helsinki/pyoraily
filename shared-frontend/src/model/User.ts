/*
 *
 * MIT License.
 *
 */

import { UserRole } from './UserRole';
import { UserStatus } from './UserStatus';

export interface User {
  id: string | null;
  firstName: string;
  lastName: string;
  gender: 'male' | 'female' | 'na';
  email: string;
  uciId?: string;
  yearOfBirth: number;
  status: UserStatus;
  createdAt?: string;
  updatedAt?: string;
  lastLogin?: string;
  userRoles: UserRole[];
}
