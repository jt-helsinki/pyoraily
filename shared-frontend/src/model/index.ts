/*
 *
 * Copyright (c) Vaisala Oyj. All rights reserved.
 *
 */

export * from './AthleteProfile';
export * from './Event';
export * from './Result';
export * from './types';
export * from './User';
export * from './UserPreferences';
export * from './UserRole';
export * from './UserStatus';
