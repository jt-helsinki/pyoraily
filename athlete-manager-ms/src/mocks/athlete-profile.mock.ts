/*
 *
 * MIT License.
 *
 */
import { AthleteProfile } from '@src/model/AthleteProfile';
import { Event } from '@src/model/Event';
import { Gender } from 'pyoraily-shared-backend/model/user/User';

export const mockAthleteProfile: AthleteProfile = {
  id: '123456789',
  userID: '123456789',
  year: 2023,
  firstName: 'John',
  lastName: 'Doe',
  yearOfBirth: 2000,
  gender: Gender.MALE,
  nominatedCategory: 'Junior',
  nominatedDisciplines: [] as string[],
  events: [] as Event[],
};
