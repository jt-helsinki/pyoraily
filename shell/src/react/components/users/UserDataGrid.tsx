/*
 *
 * MIT License.
 *
 */
import { dataGridSortComparator } from 'pyoraily-shared-frontend/utils/sortComparator';
import * as UserUtils from 'pyoraily-shared-frontend/utils/UserUtils';
import { User } from 'pyoraily-shared-frontend/model/User';
import { UserRole } from 'pyoraily-shared-frontend/model/UserRole';
import * as UserComponentUtils from '@src/react/components/common/UserComponentUtils';
import * as React from 'react';
import DataGrid, { CellRendererProps, Column, SortColumn } from 'react-data-grid';
import { Button, Dimmer, Loader } from 'semantic-ui-react';

const columns = (userRoles: UserRole[], onEditRecord: (user: User) => void): Column<User>[] => [
  {
    key: 'fullName',
    name: 'Name',
    sortable: true,
    resizable: true,
    formatter: ((record: CellRendererProps<User, null>): React.ReactNode => {
      const user: User = record.row;

      return user.firstName ? (
        <>
          {user.firstName} {user.lastName}
        </>
      ) : (
        <>Not entered</>
      );
    }) as any,
  },
  {
    key: 'email',
    name: 'Email',
    sortable: true,
    resizable: true,
    formatter: ((record: CellRendererProps<User, null>) => <span>{record.row.email || ''}</span>) as any,
  },
  {
    key: 'status',
    name: 'Status',
    resizable: true,
    formatter: ((record: CellRendererProps<User, null>): React.ReactNode => {
      const user: User = record.row;

      return <>{UserComponentUtils.userStatusToTags(user.status)}</>;
    }) as any,
  },
  {
    key: 'roles',
    name: 'Role(s)',
    sortable: false,
    resizable: true,
    formatter: ((record: CellRendererProps<User, null>): React.ReactNode => {
      const userRolesForUser = UserUtils.userRolesForUser(record.row, userRoles || []);
      return UserComponentUtils.userRolesToTags(userRolesForUser || []);
    }) as any,
  },
  {
    key: 'createdAt',
    name: 'Created',
    sortable: true,
    resizable: true,
    formatter: (record: CellRendererProps<User, null>) =>
      (<span>{new Date(record.row.createdAt as string).toLocaleDateString('en-GB')}</span>) as any,
  },
  {
    key: 'updatedAt',
    name: 'Updated',
    sortable: true,
    resizable: true,
    formatter: (record: CellRendererProps<User, null>) =>
      (<span>{new Date(record.row.updatedAt as string).toLocaleDateString('en-GB')}</span>) as any,
  },
  {
    key: 'lastLogin',
    name: 'Last Login',
    sortable: true,
    resizable: true,
    formatter: (record: CellRendererProps<User, null>) =>
      (<span>{new Date(record.row.lastLogin as string).toLocaleDateString('en-GB')}</span>) as any,
  },
  {
    key: 'edit',
    name: 'Actions',
    sortable: false,
    resizable: true,
    formatter: ((record: CellRendererProps<User, null>) => {
      const onClick = (): void => onEditRecord(record.row);
      return (
        <Button type="button" kind="ghost" size="mini" onClick={onClick}>
          Edit
        </Button>
      );
    }) as any,
  },
];

const NoRowsFallback = (props: { isFetching: boolean }): React.ReactElement =>
  props.isFetching ? (
    <div style={{ gridColumn: '1/-1' }}>Fetching users...</div>
  ) : (
    <div style={{ gridColumn: '1/-1' }}>No users found</div>
  );

interface Props {
  searchTerm: string;

  isFetching: boolean;

  userRoles: UserRole[];

  users: User[];

  onEditUser: (user: User) => void;
}

export const UserDataGrid: React.FunctionComponent<Props> = (props: Props): React.ReactElement<Props> => {
  const dataGridColumns = columns(props.userRoles || [], props.onEditUser);
  const [sortColumns, setSortColumns] = React.useState<SortColumn[]>([]);
  const sortedRows: User[] = props.users.sort(dataGridSortComparator(sortColumns));

  const filteredRows = React.useMemo(
    (): User[] =>
      props.searchTerm.length >= 3
        ? sortedRows.filter(
            (user: User) =>
              user.email?.toLocaleLowerCase().startsWith(props.searchTerm.toLocaleLowerCase()) ||
              user.lastName?.toLocaleLowerCase().startsWith(props.searchTerm.toLocaleLowerCase()) ||
              user.firstName?.toLocaleLowerCase().startsWith(props.searchTerm.toLocaleLowerCase())
          )
        : sortedRows,
    [props.searchTerm, sortedRows, sortColumns]
  );

  return (
    <>
      <Dimmer active={props.isFetching}>
        <Loader>Fetching user data...</Loader>
      </Dimmer>
      <DataGrid
        columns={dataGridColumns as any}
        rows={filteredRows}
        sortColumns={sortColumns}
        onSortColumnsChange={setSortColumns}
        defaultColumnOptions={{ width: '1fr' }}
        headerRowHeight={40}
        style={{ resize: 'both' }}
        renderers={{ noRowsFallback: <NoRowsFallback isFetching={props.isFetching} /> }}
      />
    </>
  );
};
