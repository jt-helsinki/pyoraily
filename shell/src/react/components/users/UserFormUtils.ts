/*
 *
 * MIT License.
 *
 */
import { User } from 'pyoraily-shared-frontend/model/User';
import { UserRole } from 'pyoraily-shared-frontend/model/UserRole';
import { UserStatus } from 'pyoraily-shared-frontend/model/UserStatus';
import { DropdownItemProps } from 'semantic-ui-react/dist/commonjs/modules/Dropdown/DropdownItem';

export const userRoleToMultiSelectItem = (userRole: UserRole): DropdownItemProps[] =>
  ({
    key: userRole,
    text: userRole.replace('_', ' '),
    value: userRole,
  } as any);

export const selectedUserRoleIdsToMultiSelectItems = (
  selectedUserRoleIds: string[],
  mappedUserRoles: DropdownItemProps[]
): DropdownItemProps[] => mappedUserRoles.filter((userRole) => selectedUserRoleIds.some((id) => id === userRole.id));

export const emptyUser = (): Partial<User> => ({
  id: null,
  firstName: '',
  lastName: '',
  userRoles: [],
  email: '',
  status: UserStatus.Active,
});
