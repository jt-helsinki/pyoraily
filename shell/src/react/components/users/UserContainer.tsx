/*
 *
 * MIT License.
 *
 */
import { User } from 'pyoraily-shared-frontend/model/User';
import { UserDataGrid } from '@src/react/components/users/UserDataGrid';
import * as UserFormUtils from '@src/react/components/users/UserFormUtils';
import { UserUpsertForm } from '@src/react/components/users/UserUpsertForm';
import {
  useCreateUserServerStateStore,
  useFetchUserRolesServerStateStore,
  useFetchUsersServerStateStore,
  useUpdateUserServerStateStore,
} from '@src/react/hooks/state/server/useUserServerStateStore';
import { NotificationKey } from '@src/react/types/NotificationKey';
import * as UiService from '@src/services/UiService';
import * as React from 'react';
import { Dimmer, Input, Loader, Modal } from 'semantic-ui-react';

enum ModalMode {
  None,
  EditUser,
  CreateUser,
}

export const UserContainer: React.FunctionComponent = (): React.ReactElement | null => {
  const { data: userRoles } = useFetchUserRolesServerStateStore();
  const { isFetching, data: users, refetch } = useFetchUsersServerStateStore();

  const { mutateAsync: updateUserMutate } = useUpdateUserServerStateStore();
  const { mutateAsync: createUserMutate } = useCreateUserServerStateStore();

  const [error, setError] = React.useState<string>('');
  const [isSubmitting, setIsSubmitting] = React.useState<boolean>(false);
  const [searchTerm, setSearchTerm] = React.useState<string>('');
  const [user, setUser] = React.useState<Partial<User>>(UserFormUtils.emptyUser());
  const [modalMode, setModalMode] = React.useState<ModalMode>(ModalMode.None);

  const onCancel = (): void => {
    setError(null as any);
    setIsSubmitting(false);
    setUser(UserFormUtils.emptyUser());
    setModalMode(ModalMode.None);
  };

  const handleSaveResponse = (response: Promise<User>): void => {
    setIsSubmitting(true);
    response
      .then((savedUser: User) => {
        UiService.notificationToasterSuccess(
          NotificationKey.USERS_EMAIL_UPDATED,
          `User ${savedUser.firstName} ${savedUser.lastName} <${savedUser.email}> updated.`
        );
        refetch().then(() => onCancel());
      })
      .catch((responseError: string) => {
        setError(responseError);
        setIsSubmitting(false);
      });
  };

  const onEditUser = (userToEdit: User): void => {
    setUser(userToEdit);
    setModalMode(ModalMode.EditUser);
  };

  const onEditUserSave = (userToSave: any): void => {
    handleSaveResponse(updateUserMutate(userToSave));
  };

  const visible: boolean = modalMode === ModalMode.EditUser || modalMode === ModalMode.CreateUser;

  return (
    <>
      <div>
        <p>
          <Input icon="search" placeholder="Search..." onChange={(event: any) => setSearchTerm(event.target.value)} />
        </p>
        <UserDataGrid
          searchTerm={searchTerm}
          users={users || []}
          userRoles={userRoles || []}
          isFetching={isFetching || isSubmitting}
          onEditUser={onEditUser}
        />
      </div>
      <Modal onClose={onCancel} open={visible}>
        <Dimmer active={isSubmitting}>
          <Loader>Saving</Loader>
        </Dimmer>
        {modalMode === ModalMode.EditUser ? (
          <UserUpsertForm
            isSubmitting={isSubmitting}
            title="Edit user"
            description="Change user details, permissions and roles."
            onSubmit={onEditUserSave}
            onCancel={onCancel}
            user={user as User}
            userRoles={userRoles || []}
            errorMessage={error}
          />
        ) : null}
      </Modal>
    </>
  );
};
