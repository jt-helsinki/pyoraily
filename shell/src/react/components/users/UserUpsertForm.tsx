/*
 *
 * MIT License.
 *
 */
import { UserEditFormModelSchema } from '@src/model/schema/UserFormSchemas';
import { User } from 'pyoraily-shared-frontend/model/User';
import { UserRole } from 'pyoraily-shared-frontend/model/UserRole';
import * as UserFormUtils from '@src/react/components/users/UserFormUtils';
import { Formik, FormikProps } from 'formik';
import * as React from 'react';
import { Button, Dropdown, Form, Modal, Radio } from 'semantic-ui-react';
import { DropdownItemProps } from 'semantic-ui-react/dist/commonjs/modules/Dropdown/DropdownItem';

interface Props {
  isSubmitting: boolean;

  title: string;

  description: string;

  disableNameAndEmail?: boolean;

  user: User;

  errorMessage?: string;

  onSubmit: (values: User) => void;

  onCancel: () => void;

  userRoles: UserRole[];
}

export const UserUpsertForm: React.FunctionComponent<Props> = (props: Props): React.ReactElement<Props> | null => {
  const mappedUserRoles = React.useMemo(
    (): DropdownItemProps[] => props.userRoles.map(UserFormUtils.userRoleToMultiSelectItem),
    [props.userRoles]
  );
  return (
    <>
      <Modal.Header>{props.title}</Modal.Header>
      <Modal.Content>
        <Modal.Description>{props.description}</Modal.Description>
        <Formik
          validateOnChange={true}
          validateOnBlur={true}
          enableReinitialize={true}
          initialValues={props.user}
          validationSchema={UserEditFormModelSchema}
          onSubmit={(values: User): void => {
            props.onSubmit(values);
          }}>
          {(formik: FormikProps<User>): React.ReactElement => (
            <Form id="user-upsert-form">
              <Form.Input
                type="text"
                required={true}
                id="user-upsert-form-first-name"
                name="firstName"
                label="First name"
                value={formik.values.firstName}
                disabled={props.isSubmitting}
                error={formik.errors.firstName}
                onChange={formik.handleChange}
              />
              <Form.Input
                type="text"
                required={true}
                id="user-upsert-form-last-name"
                name="lastName"
                label="Last name"
                value={formik.values.lastName}
                disabled={props.isSubmitting}
                error={formik.errors.lastName}
                onChange={formik.handleChange}
              />
              <Form.Input
                type="text"
                required={true}
                id="user-upsert-form-email"
                name="email"
                label="Email"
                value={formik.values.email}
                disabled={props.isSubmitting}
                error={formik.errors.email}
                onChange={formik.handleChange}
              />
              <Form.Field>
                <label>Roles</label>
                <Dropdown
                  id="user-upsert-form-user-roles"
                  name="userRoles"
                  placeholder="Roles"
                  fluid={true}
                  multiple={true}
                  selection
                  options={mappedUserRoles}
                  disabled={props.isSubmitting}
                  defaultValue={formik.values.userRoles}
                  error={formik.errors.userRoles as any}
                  onChange={(event: any, data: any) => formik.setFieldValue('userRoles', data.value, false)}
                />
              </Form.Field>
              <Form.Group inline={true} unstackable={false}>
                <label>Status</label>
                <Form.Field
                  control={Radio}
                  label="Active"
                  value="Active"
                  name="status"
                  disabled={props.isSubmitting}
                  checked={formik.values.status === 'Active'}
                  onClick={() => formik.setFieldValue('status', 'Active', false)}
                />
                <Form.Field
                  control={Radio}
                  label="Blocked"
                  value="Blocked"
                  name="status"
                  disabled={props.isSubmitting}
                  checked={formik.values.status === 'Blocked'}
                  onClick={() => formik.setFieldValue('status', 'Blocked', false)}
                />
              </Form.Group>
              <Button secondary={true} onClick={props.onCancel}>
                Cancel
              </Button>
              <Button primary={true} onClick={formik.submitForm}>
                Save
              </Button>
            </Form>
          )}
        </Formik>
      </Modal.Content>
    </>
  );
};

UserUpsertForm.defaultProps = {
  disableNameAndEmail: false,
};
