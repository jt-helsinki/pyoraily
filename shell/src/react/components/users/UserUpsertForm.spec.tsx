import { User } from 'pyoraily-shared-frontend/model/User';
import { UserUpsertForm } from '@src/react/components/users/UserUpsertForm';
import { MOCK_USER_ROLES } from '@src/tests/mocks/mockUserRoleServiceResponses';
import { mockFetchProfileResponseDate } from '@src/tests/mocks/mockUserServiceResponses';
import { getById, renderSUI } from '@src/tests/utils';
import { act, fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

(global as any).IS_REACT_ACT_ENVIRONMENT = true;

describe('UserEditForm', () => {
  let user: User;
  let emptyUser: User;

  beforeEach(() => {
    (global as any).IS_REACT_ACT_ENVIRONMENT = true;
    user = mockFetchProfileResponseDate;
    emptyUser = {
      firstName: '',
    } as User;
  });

  it('should match the snapshot', () => {
    const { container } = renderSUI(
      <UserUpsertForm
        title="Test title"
        description="Test description"
        disableNameAndEmail={false}
        isSubmitting={false}
        onSubmit={() => {}}
        onCancel={() => {}}
        user={user}
        userRoles={MOCK_USER_ROLES}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should display user form title and description', () => {
    const { container } = renderSUI(
      <UserUpsertForm
        title="Test title"
        description="Test description"
        disableNameAndEmail={false}
        isSubmitting={false}
        onSubmit={() => {}}
        onCancel={() => {}}
        user={user}
        userRoles={MOCK_USER_ROLES}
      />
    );
    expect(screen.getByText('Test title')).toBeInTheDocument();
    expect(screen.getByText('Test description')).toBeInTheDocument();
    expect(screen.getByText('Save')).toBeInTheDocument();
    expect(screen.getByText('Cancel')).toBeInTheDocument();
  });
  //
  // it('should disable save button', () => {
  //   renderSUI(
  //     <UserUpsertForm
  //       title="Test title"
  //       description="Test description"
  //       disableNameAndEmail={false}
  //       isSubmitting={false}
  //       visible={true}
  //       onSubmit={() => {}}
  //       onCancel={() => {}}
  //       user={emptyUserEditFormData}
  //       userRoles={[]}
  //       userRoles={[]}
  //     />
  //   );
  //   expect(screen.getByRole('button', { name: 'Save' })).toBeDisabled();
  // });
  //
  it('should render initial user data', () => {
    const { container } = renderSUI(
      <UserUpsertForm
        title="Test title"
        description="Test description"
        disableNameAndEmail={false}
        isSubmitting={false}
        onSubmit={() => {}}
        onCancel={() => {}}
        user={user}
        userRoles={MOCK_USER_ROLES}
      />
    );
    expect(getById(container, 'user-upsert-form-first-name')).toHaveValue('Skid');
    expect(getById(container, 'user-upsert-form-last-name')).toHaveValue('Mark');
    expect(getById(container, 'user-upsert-form-email')).toHaveValue('test@test.com');
  });

  it('should show an error message for the name', async () => {
    const user = userEvent.setup();
    const { container } = renderSUI(
      <UserUpsertForm
        title="Test title"
        description="Test description"
        disableNameAndEmail={false}
        isSubmitting={false}
        onSubmit={() => {}}
        onCancel={() => {}}
        user={emptyUser}
        userRoles={MOCK_USER_ROLES}
      />
    );

    const input = getById(container, 'user-upsert-form-first-name');

    expect(input).toHaveValue('');
    expect(screen.queryByText('First name is required')).not.toBeInTheDocument();

    await user.click(input);
    expect(screen.queryByText('First name is required')).not.toBeInTheDocument();

    await user.tab();
    // expect(screen.queryByText('Name is required')).toBeInTheDocument();
    // expect(screen.queryByText('Email is required')).not.toBeInTheDocument();
    //
    // user.type(input, 'A');
    // expect(screen.queryByText('Name is too short')).toBeInTheDocument();
  });

  it('should show an error message for the email', async () => {
    const user = userEvent.setup();
    const { container } = renderSUI(
      <UserUpsertForm
        title="Test title"
        description="Test description"
        disableNameAndEmail={false}
        isSubmitting={false}
        onSubmit={() => {}}
        onCancel={() => {}}
        user={emptyUser}
        userRoles={MOCK_USER_ROLES}
      />
    );
    const input = getById(container, 'user-upsert-form-email');

    expect(input).toHaveValue('');
    expect(screen.queryByText('Email is required')).not.toBeInTheDocument();

    await user.click(input);
    expect(screen.queryByText('Email is required')).not.toBeInTheDocument();

    await user.tab();
    // expect(screen.queryByText('Email is required')).toBeInTheDocument();
    // expect(screen.queryByText('Name is required')).not.toBeInTheDocument();
    //
    // await user.type(input, 'sink@hole.com');
    // expect(screen.queryByText('Email is required')).not.toBeInTheDocument();
  });

  it('should show an error message when the name is too short', async () => {
    const user = userEvent.setup();
    const { container } = renderSUI(
      <UserUpsertForm
        title="Test title"
        description="Test description"
        disableNameAndEmail={false}
        isSubmitting={false}
        onSubmit={() => {}}
        onCancel={() => {}}
        user={emptyUser}
        userRoles={MOCK_USER_ROLES}
      />
    );
    const input = getById(container, 'user-upsert-form-first-name');

    await act(async () => {
      expect(input).toHaveValue('');
      expect(screen.queryByText('Name is required')).not.toBeInTheDocument();

      await user.click(input);
      expect(screen.queryByText('Name is required')).not.toBeInTheDocument();
      expect(screen.queryByText('Name is too short')).not.toBeInTheDocument();

      // expect(screen.queryByText('Name is too short')).toBeInTheDocument();
      //
      // await user.type(input, 'A'); // now value is AA
      // expect(screen.queryByText('Name is too short')).not.toBeInTheDocument();
    });

    await act(async () => {
      await user.type(input, 'A');
      expect(input).toHaveValue('A');
      await fireEvent.blur(input);
    });
  });

  // it('should show an error message: name is too short', async () => {
  //   renderSUI(
  //     <UserUpsertForm
  //       title="Test title"
  //       description="Test description"
  //       disableNameAndEmail={false}
  //       isSubmitting={false}
  //       visible={true}
  //       onSubmit={() => {}}
  //       onCancel={() => {}}
  //       user={emptyUserEditFormData}
  //       userRoles={[]}
  //     />
  //   );
  //   const nameInput = screen.getByRole('textbox', { name: 'name' });
  //   fireEvent.change(nameInput, { target: { value: 'a' } });
  //   fireEvent.blur(nameInput);
  //   await waitFor(() => {
  //     expect(screen.getByText('Name is too short')).toBeInTheDocument();
  //   });
  // });
  //
  // it('should show an error message: name is too long', async () => {
  //   renderSUI(
  //     <UserUpsertForm
  //       title="Test title"
  //       description="Test description"
  //       disableNameAndEmail={false}
  //       isSubmitting={false}
  //       visible={true}
  //       onSubmit={() => {}}
  //       onCancel={() => {}}
  //       user={emptyUserEditFormData}
  //       userRoles={[]}
  //     />
  //   );
  //   const nameInput = screen.getByRole('textbox', { name: 'name' });
  //   const value = 'a'.repeat(130);
  //   fireEvent.change(nameInput, { target: { value } });
  //   fireEvent.blur(nameInput);
  //   await waitFor(() => {
  //     expect(screen.getByText('Name is too long')).toBeInTheDocument();
  //   });
  // });
  //
  // it('should show an error messag: invalid name', async () => {
  //   renderSUI(
  //     <UserUpsertForm
  //       title="Test title"
  //       description="Test description"
  //       disableNameAndEmail={false}
  //       isSubmitting={false}
  //       visible={true}
  //       onSubmit={() => {}}
  //       onCancel={() => {}}
  //       user={emptyUserEditFormData}
  //       userRoles={[]}
  //     />
  //   );
  //   const nameInput = screen.getByRole('textbox', { name: 'name' });
  //   fireEvent.change(nameInput, { target: { value: '123' } });
  //   fireEvent.blur(nameInput);
  //   await waitFor(() => {
  //     expect(screen.getByText('Invalid name')).toBeInTheDocument();
  //   });
  // });
  //
  // it('should show an error messag: email is required', async () => {
  //   renderSUI(
  //     <UserUpsertForm
  //       title="Test title"
  //       description="Test description"
  //       disableNameAndEmail={false}
  //       isSubmitting={false}
  //       visible={true}
  //       onSubmit={() => {}}
  //       onCancel={() => {}}
  //       user={emptyUserEditFormData}
  //       userRoles={[]}
  //     />
  //   );
  //   const emailInput = screen.getByRole('textbox', { name: 'email' });
  //   fireEvent.change(emailInput, { target: { value: '' } });
  //   fireEvent.blur(emailInput);
  //   await waitFor(() => {
  //     expect(screen.getByText('Email is required')).toBeInTheDocument();
  //   });
  // });
  //
  // it('should show an error messag: invalid email', async () => {
  //   renderSUI(
  //     <UserUpsertForm
  //       title="Test title"
  //       description="Test description"
  //       disableNameAndEmail={false}
  //       isSubmitting={false}
  //       visible={true}
  //       onSubmit={() => {}}
  //       onCancel={() => {}}
  //       user={emptyUserEditFormData}
  //       userRoles={[]}
  //     />
  //   );
  //   const emailInput = screen.getByRole('textbox', { name: 'email' });
  //   fireEvent.change(emailInput, { target: { value: 'test.user' } });
  //   fireEvent.blur(emailInput);
  //   await waitFor(() => {
  //     expect(screen.getByText('Invalid email')).toBeInTheDocument();
  //   });
  // });
  //
  // it('should call onSubmit when save button is clicked', async () => {
  //   const mockSubmit = jest.fn();
  //   renderSUI(
  //     <UserUpsertForm
  //       disableNameAndEmail={false}
  //       isSubmitting={false}
  //       visible={true}
  //       title="Test title"
  //       description="Test description"
  //       onSubmit={(values) => mockSubmit(values)}
  //       onCancel={() => {}}
  //       user={emptyUserEditFormData}
  //       userRoles={[]}
  //     />
  //   );
  //   const nameInput = screen.getByRole('textbox', { name: 'name' });
  //   fireEvent.change(nameInput, { target: { value: 'Test user' } });
  //   const emailInput = screen.getByRole('textbox', { name: 'email' });
  //   fireEvent.change(emailInput, {
  //     target: { value: 'test.user@example.com' },
  //   });
  //   const saveButton = screen.getByRole('button', { name: 'Save' });
  //   fireEvent.click(saveButton);
  //   await waitFor(() => {
  //     expect(mockSubmit).toHaveBeenCalledWith({
  //       name: 'Test user',
  //       email: 'test.user@example.com',
  //       roles: [],
  //       blocked: false,
  //     });
  //   });
  // });
});
