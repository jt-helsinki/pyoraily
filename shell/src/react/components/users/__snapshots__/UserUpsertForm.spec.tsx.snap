// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserEditForm should match the snapshot 1`] = `
<div>
  <div
    class="header"
  >
    Test title
  </div>
  <div
    class="content"
  >
    <div
      class="description"
    >
      Test description
    </div>
    <form
      class="ui form"
      id="user-upsert-form"
    >
      <div
        class="required field"
      >
        <label
          for="user-upsert-form-first-name"
        >
          First name
        </label>
        <div
          class="ui input"
        >
          <input
            id="user-upsert-form-first-name"
            name="firstName"
            required=""
            type="text"
            value="Skid"
          />
        </div>
      </div>
      <div
        class="required field"
      >
        <label
          for="user-upsert-form-last-name"
        >
          Last name
        </label>
        <div
          class="ui input"
        >
          <input
            id="user-upsert-form-last-name"
            name="lastName"
            required=""
            type="text"
            value="Mark"
          />
        </div>
      </div>
      <div
        class="required field"
      >
        <label
          for="user-upsert-form-email"
        >
          Email
        </label>
        <div
          class="ui input"
        >
          <input
            id="user-upsert-form-email"
            name="email"
            required=""
            type="text"
            value="test@test.com"
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label>
          Roles
        </label>
        <div
          aria-disabled="false"
          aria-expanded="false"
          aria-multiselectable="true"
          class="ui fluid multiple selection dropdown"
          id="user-upsert-form-user-roles"
          name="userRoles"
          role="listbox"
          tabindex="0"
        >
          <a
            class="ui label"
            value="HPY"
          >
            HPY
            <i
              aria-hidden="true"
              class="delete icon"
            />
          </a>
          <a
            class="ui label"
            value="ATHLETE"
          >
            ATHLETE
            <i
              aria-hidden="true"
              class="delete icon"
            />
          </a>
          <a
            class="ui label"
            value="SP"
          >
            SP
            <i
              aria-hidden="true"
              class="delete icon"
            />
          </a>
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider text"
            role="alert"
          >
            Roles
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="false"
              aria-selected="true"
              class="selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                ADMIN
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="inline fields"
      >
        <label>
          Status
        </label>
        <div
          class="field"
        >
          <div
            class="ui checked radio checkbox"
          >
            <input
              checked=""
              class="hidden"
              name="status"
              readonly=""
              tabindex="0"
              type="radio"
              value="Active"
            />
            <label>
              Active
            </label>
          </div>
        </div>
        <div
          class="field"
        >
          <div
            class="ui radio checkbox"
          >
            <input
              class="hidden"
              name="status"
              readonly=""
              tabindex="0"
              type="radio"
              value="Blocked"
            />
            <label>
              Blocked
            </label>
          </div>
        </div>
      </div>
      <button
        class="ui secondary button"
      >
        Cancel
      </button>
      <button
        class="ui primary button"
      >
        Save
      </button>
    </form>
  </div>
</div>
`;
